Add initial version of Python bindings

RalfG opened the pull request:
A first implementation of Python bindings for the rust library. I started out with objects that seemed to make most sense to expose, but more can be added. Additionally, for some objects, more fields could be exposed in the future.

I thought it would be good to keep all code related to the Python bindings separate from the main rust package itself. That does mean that all object have to be wrapped with a "Python" version, resulting in a lot of boilerplate code. 

The Python package is built automatically on GitHub Actions for each PR or push to `release`. On a tag (or GitHub release), the package is uploaded to PyPI. For this, [trusted publishing](https://packaging.python.org/en/latest/guides/publishing-package-distribution-releases-using-github-actions-ci-cd-workflows/#configuring-trusted-publishing) should still be set up.

In terms of version management, I thought it would be good to keep in sync with the rust library.

Documentation is built automatically with Sphinx on readthedocs.org. A version is already online from my branch: https://rustyms.readthedocs.io.

RalfG committed:
initial commit of python bindings

RalfG committed:
Fix readthedocs config extension

RalfG committed:
Add cargo install for readthedocs

RalfG committed:
Use . instead of source

RalfG committed:
Use builtin rust tool config

douweschulte commented on docs/python/source/conf.py:
```diff
@@ -0,0 +1,53 @@
+# Project information
+project = "rustyms"
+author = "Douwe Schulte, Ralf Gabriels"
+github_project_url = "https://github.com/compomics/rustyms/"
```
At some point these can be updated ;-)

RalfG replied:
Whoops. `compomics` was a copy-paste error from another project.

douweschulte commented on rustyms-py/Cargo.toml:
```diff
@@ -0,0 +1,13 @@
+[package]
+name = "rustyms_py"
```
I have always used 'kebab case' for package names (meaning I would use `rustyms-py` here.

RalfG replied:
Fixed!

douweschulte reviewed (commented):
Looks really nice! You seem to already have set up a lot of things, very nice work.

RalfG committed:
Update python docs

RalfG committed:
Update branch in python ci

RalfG committed:
Implement requested changes

RalfG committed:
Minor fix
Pass error instead of using blind unwrap.

douweschulte commented on .vscode/settings.json:
```diff
@@ -107,5 +109,9 @@
+    "rust-analyzer.server.extraEnv": {
+        "PYO3_PYTHON": "./.venv/Scripts/python.exe"
+    },
+    "rust-analyzer.linkedProjects": ["Cargo.toml", "rustyms-py/Cargo.toml"]
```
Tiny question as it was added as a project in the root Cargo.toml, do we still need this setting set up?

RalfG replied:
Not sure. I get this error without it:

```
ERROR rust_analyzer::main_loop: FetchWorkspaceError:
rust-analyzer failed to discover workspace
```

For now, I removed it from git again.

douweschulte replied:
If you get an error without then it is good to leave it in.

douweschulte commented on docs/python/source/index.md:
```diff
@@ -0,0 +1,72 @@
+- {py:class}`~rustyms.SequenceElement`: One position in a peptide sequence with amino acid and
+  modifications
+- {py:class}`~rustyms.LinearPeptide`: Peptide sequence, modifications, and charge, with full
+  [ProForma 2.0](https://proforma.readthedocs.io) support
```
It does not have full ProForma support, I want to get there eventually but did not get around to it yet. I would send people to the `pro_forma_grammar.md` file in the root for details on what exactly is supported.

douweschulte commented on docs/python/source/index.md:
```diff
@@ -0,0 +1,72 @@
+- {py:class}`~rustyms.SequenceElement`: One position in a peptide sequence with amino acid and
+  modifications
+- {py:class}`~rustyms.LinearPeptide`: Peptide sequence, modifications, and charge, with full
+  [ProForma 2.0](https://proforma.readthedocs.io) support
```
The url for the pro forma spec does not link to anything. https://github.com/HUPO-PSI/ProForma seems like a better url to me or potentially https://www.psidev.info/proforma.

RalfG replied:
I am working on proforma.readthedocs.io in parallel to this. It should come online sometime this week.

douweschulte commented on rustyms-py/CONTRIBUTING.md:
````diff
@@ -0,0 +1,52 @@
+pip install --editable ./rustyms-py
+```
+
+## Documentation
````
Once this is all set up I would like to have the link to the docs somewhere in the main readme as well.

RalfG replied:
Added an entry to the main readme.

douweschulte commented on rustyms-py/README.md:
```diff
@@ -0,0 +1,14 @@
+# Python bindings to rustyms
+
+Python bindings to the [rustyms](https://docs.rs/rustyms/0.7.0/rustyms/) library for proteomics
```
This url contains a specific version.

douweschulte commented on rustyms-py/src/lib.rs:
```diff
@@ -0,0 +1,1262 @@
+    /// -------
+    /// float | None
+    ///
+    fn most_abundant_mass(&self) -> Option<f64> {
```
This one is broken upstream as well, so maybe remove this from the public api for now (I should maybe to the same on the rust side)

douweschulte commented on rustyms-py/src/lib.rs:
```diff
@@ -0,0 +1,1262 @@
+    ///
+    fn most_abundant_mass(&self) -> Option<f64> {
+        match self.0.most_abundant_mass() {
+            Some(mass) => Some(mass.value),
```
Maybe a bit pedantic but this looks like a manual map implementation `self.0.most_abundant_mass().map(|m| m.value)` does the same, normally clippy will highlight these cases to you, did you run clippy over it already?

RalfG replied:
Thanks for the tip! I'm still pretty new to Rust, so similar idiomatic mistakes might pop up more.

douweschulte commented on rustyms-py/src/lib.rs:
```diff
@@ -0,0 +1,1262 @@
+    /// ValueError
+    ///   If the mode is not one of the valid modes.
+    ///
+    fn mass(&self, mode: &str) -> PyResult<Option<f64>> {
```
I have seen this in the model as well, but having things stringly typed (using strings that have to be of a specific format) is frowned upon. Is there a way to set up an enum at the python side that contains these three possible cases?

RalfG replied:
Fixed it by making a pyclass Enum.

douweschulte commented on rustyms-py/src/lib.rs:
```diff
@@ -0,0 +1,1262 @@
+    /// -------
+    /// MolecularFormula
+    ///
+    fn formula(&self, selected_ambiguous: usize) -> MolecularFormula {
```
Did you also provide methods for the `formulas` equivalents? Those are the only ones I use internally, with these exposed for people that know their stuff does not contain B/Z.

RalfG replied:
I'm not entirely following here. What do you mean with `formulas` equivalents?

douweschulte replied:
The methods called `formulas` that returns a `Multi<Mass>`.  There are `*_greedy` and `*_all` equivalents.

RalfG replied:
Whoops. Was not using the latest rustyms from the repo, but using 0.7.0 instead. Had not yet noticed all the changes.

douweschulte commented on rustyms-py/src/lib.rs:
```diff
@@ -0,0 +1,1262 @@
+        self.0.sequence.iter().map(|x| x.aminoacid.char()).collect()
+    }
+
+    // TODO: How to get the charge as an integer?
```
`self.formulas().next().unwrap().charge()` This takes the first formula and get the charge out of that, it unwraps because it assumes there to be at least one formula (which is always the cases) and only takes the first as the only difference will be in the count of C/H due to B/Z not in the actual charge. But I did not provide any method like this on the Rust side, if you think this is a useful method to have regardless I will set it up on the rust side as well.

I did not set it up I think because I never really regard any peptide as intrinsically charged, I just take it as a sequence and overwrite the charges in the generate_fragments method.

RalfG replied:
Thanks!

It's indeed a bit strange to assign a charge in a ProForma peptidoform sequence, as it is a property of a precursor, not of a peptidoform. Nevertheless, it was added out of convenience for bottom up MS. For instance, I deal a lot with peptide spectrum MS2 intensities, which are mostly defined by the combination peptide-modifications-charge. 

To me it would make sense to also have direct access to the charge state on the rust side.

douweschulte replied:
Good I will think about the cleanest way to implement it on the Rust side then, then this function can become a direct port again.

RalfG replied:
Could it be that the charge is currently not parsed at all when creating a peptide with `ComplexPeptide::pro_forma`?

douweschulte replied:
There is a known limitation in my parser where it fails to parse c term modifications followed by a charge state and a chimeric peptide. I will have to address that at some point. But if you see something else that is going awry please let me know then I can add some more tests and fix it.

RalfG replied:
It works now. Before I tried to get the charge from `formula()` instead of `charge_carriers().formula()`.

douweschulte commented on rustyms-py/src/lib.rs:
```diff
@@ -0,0 +1,1262 @@
+        &self,
+        max_charge: i16,
+        model: &str,
+        // peptide_index: usize, TODO: Required for linear peptide?
```
It is required for a linear one because the complex calls exactly the same function. Maybe it would be nicer indeed to write some in between method that auto fills 0 in the linear peptide case.

douweschulte commented on rustyms-py/src/lib.rs:
```diff
@@ -0,0 +1,1262 @@
+                &model,
+                0, // TODO: Don't hard code?
+            )
+            .unwrap()
```
At the rust side this function should not be able to crash anymore (now that I build support for B/Z).

RalfG replied:
Should I replace the `unwrap` with something else?

douweschulte replied:
No the result of this function is just a list of fragments it is not a `Result` anymore, so Rust should just give a compilation error.

douweschulte commented on rustyms-py/src/lib.rs:
```diff
@@ -0,0 +1,1262 @@
+        )
+    }
+
+    /// The charge of the peak.
```
This is a bit of a weird field, it is present in some mgf files. But personally I always disregard this and it is never used internally. So at least that would be nice to include in the documentation.

RalfG replied:
I would also opt to remove it entirely from `RawPeak`. It makes sense to have a charge assigned to an annotated peak, but not here.

douweschulte replied:
I assume this has some use in deconvoluted spectra, but for now we can remove it from the python bindings, removing it from the Rust side is a breaking change and I do want a bit of a makeover on the data input side anyways.

douweschulte commented on rustyms-py/src/lib.rs:
```diff
@@ -0,0 +1,1262 @@
+        self.0.intensity
+    }
+
+    /// The charge of the peak.
```
The same (about charge in RawPeak) is true for this charge, so technically this could even be different from the charge in the annotation(s).

douweschulte commented on rustyms-py/src/lib.rs:
```diff
@@ -0,0 +1,1262 @@
+        self.0.charge.value as i16
+    }
+
+    /// The annotation of the peak, if present.
```
This is a list, so I would say: "all annotations of this peaks (there can be none)".

douweschulte commented on rustyms-py/src/lib.rs:
```diff
@@ -0,0 +1,1262 @@
+    }
+
+    // fn __str__(&self) -> String {
+    //     todo!()
```
Maybe the simplest would be to use Debug. Otherwise this could be left empty until I build MGF export at some point.

douweschulte reviewed (commented):
Very nice work! I do miss however the bindings for `ComplexPeptide`, I would like to see them added in the long run (could be in a different PR though) as that encapsulates some more of the hard ProForma 2 cases. 

Do you have anything else in mind that needs to happen before merging?

douweschulte commented:
I am sorry I did a bit of restructuring again. I did not like the idea of having to set up 6 crates whenever I want to publish a new release. The API will not have changed though except for two minor things:
* `LinearPeptide::generate_theoretical_fragments` now does not take the `peptide_index` anymore
* `Species::from_str` now returns a full error with names that might be what you intended and it takes any capitalization

All in all this 'should' not result in any major work for you in this PR.

RalfG committed:
Undo changes to .vscode/settings.json

RalfG committed:
Merge branch 'release' of https://github.com/snijderlab/rustyms into release

RalfG committed:
Implement requested changes (part 1, docs)

RalfG committed:
Fix sdist build job

RalfG committed:
Implement requested changes (part 2, code, clippy)

RalfG committed:
Implement requested changes (part 3); update with upstream changes

RalfG committed:
Fix sdist build

RalfG committed:
Do not expose peak charge fields

RalfG committed:
Fix getting charge of LinearPeptide

RalfG commented:
Current to do's in the module:

- [x] `Element.most_abundant_mass()`: ~Waiting on upstream fix, or could be removed~ Removed
- [x] `MolecularFormula`: ~Can this be constructed from a string?~ Done
- [x] `MolecularFormula.with_global_isotope_modifications()`: ~Could not get the mapping to Python types working so far.~ Fixed
- [x] `MolecularFormula.most_abundant_mass()`: Idem as above
- [x] `MolecularCharge`: ~Are you planning to make this public? If not, I cannot expose it in Python. Currently, peptide charge can be passed through, but not the adduct information (I think).~ Implemented
- [ ] `FragmentType`: Not yet mapped to Python, so returned as strings.
- [x] `RawSpectrum`: (re)implement constructor

douweschulte commented:
Here are come comments on your todo's

- [ ] The `most_abundant_mass` is indeed broken upstream, for now keep this private in the Python api. I will have to dive into isotope maths to fix this at some point, which is something I want to do but does not have a very high priority.
- [x] Molecular formulas can be constructed from strings, using the `ProForma` notation and the `PSI-MOD` notation. I am not sure though if these are public members. I will check and add them if necessary.
- [ ] The `with_global_isotope` it looks like a vector should be able to work but I have never worked with PyO3 before.
- [x] `MolecularCharge` should have been public, oops.

RalfG committed:
Update dependencies

douweschulte commented:
I added a charge method to `MultiFormula` (the trait), publicised `MolecularCharge`, and added two methods fro constructing a `MolecularFormula` from string (Pro Forma and PSI-MOD, Unimod has a more advanced systems that also incorporates glycans). Additionally for the spectrum generation, I would go and use `RawSpectrum::default()` followed by `RawSpectrum::extend()` to create a new spectrum.

So pretty much done! Very nice work, is there something you still want to implement in the PR or do you want it merged as soon as possible?

RalfG committed:
Merge branch 'release' of https://github.com/snijderlab/rustyms into release

RalfG committed:
Add `MolecularCharge` binding; add `MolecularFormula.with_global_isotope_modifications` function; Remove `most_abundant_mass` bindings

Note: `MolecularCharge.charge_carriers` had to be made a public field. This can be reverted if undesired.

RalfG committed:
Reimplement RawSpectrum constructor

RalfG committed:
Implement MolecularFormula from ProForma or PSI-MOD notation

RalfG committed:
Bump version number

RalfG committed:
Configure CI for trusted publishing to PyPI

RalfG commented:
All remaining things I wanted to implement are done, so I'm good for merging. 

For the publishing of the Python package, the workflow should run on any tagged release. Not sure how you handle releases currently?

RalfG commented:
Also note that to implement bindings to `MolecularCharge`, I had to make `charge_carriers` a public field. This can be reverted if you do not want this.

douweschulte commented on rustyms-py/Cargo.toml:
```diff
@@ -0,0 +1,14 @@
+
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
+[lib]
+name = "rustyms_py"
```
Does it make sense to have kebab case here as well? Or was this underscore needed?

RalfG replied:
I don't think it's needed. Probably a left-over from my initial setup (still with Python naming conventions in mind).

RalfG replied:
Apparently it is needed:

```
rust-analyzer failed to load workspace: Failed to load the project at C:\Users\ralfg\git\rustyms\rustyms-py/Cargo.toml: Failed to read Cargo metadata from Cargo.toml file C:\Users\ralfg\git\rustyms\rustyms-py/Cargo.toml, Some(Version { major: 1, minor: 76, patch: 0 }): Failed to run `"cargo" "metadata" "--format-version" "1" "--manifest-path" "C:\\Users\\ralfg\\git\\rustyms\\rustyms-py/Cargo.toml" "--filter-platform" "x86_64-pc-windows-msvc"`: `cargo metadata` exited with an error: error: failed to parse manifest at `C:\Users\ralfg\git\rustyms\rustyms-py\Cargo.toml`

Caused by:
  library target names cannot contain hyphens: rustyms-py
```

douweschulte replied:
Alright with me then.

douweschulte commented:
Release right now are handled by locally running `cargo publish` and then once that is done tagging the correct commit in github. There should be some ways to automate this and make it do this stuff automatically. But those things are a bit of pain to set up and I did not bother yet. Also I have been fighting with many of the releases as the included databases can go over the 10 MB max file upload for crates.io very easily.

RalfG committed:
Formatting

RalfG commented:
Ready to merge.

If everything goes well, the Python package should be build on GitHub Actions and pushed to PyPI whenever a tag is pushed to GitHub.

douweschulte commented:
That sounds splendid. Is there any more setup to be done from my side? (Any secrets I should add or something?)

douweschulte commented:
We can try to push out 0.8.2 when everything has merged, this would contain the small additions I made for this PR and could be a very good test case to see if everything is running.
